README.md
# Recent custom posts

**INACTIVE NOTICE: This plugin is unsupported by WPMUDEV, we've published it here for those technical types who might want to fork and maintain it for their needs.**

## Translations

Translation files can be found at https://github.com/wpmudev/translations

## Recent custom posts lets you display a list of custom posts on any widgetized area of your site.

It's easy to add recent posts, but what about custom post types? This plugin adds a widget to your site for showing recent posts from any post type.

![Recent Custom Posts](http://premium.wpmudev.org/wp-content/uploads/2011/05/post-types.png)

 Display your latest products, classifieds ads or any other custom post types.

### Simple customizable widget

Recent Custom Posts works with any custom post type, even post types created by another plugin.

##### Customization options include:

*   Widget title
*   Post type, including core post, page and attachment types
*   Up to 20 posts
*   Include your featured image as a thumbnail for each post

Plus, you can create as many instances of the widget as you need for different custom post types.

### Integrate with other WPMU products

Publish links to custom posts created by other WPMU DEV products. Display ads created by [Classifieds](http://premium.wpmudev.org/project/classifieds), or eCommerce products using [MarketPress](http://premium.wpmudev.org/project/e-commerce "MarketPress"). Tell visitors how great you are and display quotes from happy customers using [Testimonials](http://www.travisballard.com/wordpress/tb-testimonials/) or display [Wiki pages](http://premium.wpmudev.org/project/wordpress-wiki), [Forum topics](http://justintadlock.com/archives/2010/07/16/a-wordpress-forum-plugin-using-custom-post-types) and more. Extend the way you share custom post types on your WordPress site.

## Usage

For help with installing plugins please see our [Plugin installation guide](https://premium.wpmudev.org/wpmu-manual/installing-regular-plugins-on-wpmu/). Once installed activate the plugin:

*   On regular WordPress installs - visit **Plugins** and Activate the plugin.
*   For WordPress Multisite installs - Activate it site-by-site basis (say if you wanted to make it a Pro Sites premium plugin), or visit **Network Admin -> Plugins** and Network Activate the plugin.

### To use:

1\. Go to **Appearance -> Widgets** in the site admin dashboard. 2\. Drag the new Recent Custom Posts Type Widget to your sidebar. 

![image](https://premium.wpmudev.org/wp-content/uploads/2011/05/customwidget63.jpg)

 3\. Configure the widget, selecting your custom post type. 

![image](https://premium.wpmudev.org/wp-content/uploads/2011/05/customwidget61.jpg)

 4\. Visit your home page to view the new widget in action!


changelog.txt
Plugin Name: Recent Custom Posts Type Widget
Author: Incsub

Change Log:
----------------------------------------------------------------------

2.1.3 - 2016-12-14

*Fixed:
    - Deprecated constructor warning


2.1.2 - 2013-03-13
----------------------------------------------------------------------
- Custom loop clearing.
- Removing references.

2.1.1 - 2012-12-21
----------------------------------------------------------------------
- PHP<5.3 compatibility layer (thanks, Arnold).

2.1 - 2012-10-26
----------------------------------------------------------------------
- Fixed permalink order issue.
- Added post author limit option.
- Added featured image flag.

2.0 - 2012-09-11
----------------------------------------------------------------------
- Improved admin and list output.
- Added support for more common output fields.
- Added custom fields support.

1.2 - 2011-05-16
----------------------------------------------------------------------
- Added widget option for toggling post dates on/off.
- Added order by/order direction widget options.

1.0.1 - 2011-05-14
----------------------------------------------------------------------
- Added the optional CSS class argument to widget settings.

1
----------------------------------------------------------------------
- Initial release

recent-custom-post-type.php
<?php
/*
Plugin Name: Recent Custom Posts Type Widget
Plugin URI: http://premium.wpmudev.org/project/recent-custom-posts-type-widget
Description: Allows you to display recent custom post types.
Version: 2.1.3
Author: WPMU DEV
Author URI: http://premium.wpmudev.org
WDP ID: 226

Copyright 2009-2011 Incsub (http://incsub.com)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License (Version 2 - GPLv2) as published by
the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/


///////////////////////////////////////////////////////////////////////////
/* -------------------- Update Notifications Notice -------------------- */
if ( ! function_exists( 'wdp_un_check' ) ) {
	add_action( 'admin_notices', 'wdp_un_check', 5 );
	add_action( 'network_admin_notices', 'wdp_un_check', 5 );
	function wdp_un_check() {
		if ( ! class_exists( 'WPMUDEV_Update_Notifications' ) && current_user_can( 'install_plugins' ) ) {
			echo '<div class="error fade"><p>' . __( 'Please install the latest version of <a href="http://premium.wpmudev.org/project/update-notifications/" title="Download Now &raquo;">our free Update Notifications plugin</a> which helps you stay up-to-date with the most stable, secure versions of WPMU DEV themes and plugins. <a href="http://premium.wpmudev.org/wpmu-dev/update-notifications-plugin-information/">More information &raquo;</a>', 'wpmudev' ) . '</a></p></div>';
		}
	}
}

/* --------------------------------------------------------------------- */


class RcptWidget extends WP_Widget {

	var $_order_options = array();
	var $_order_directions = array();

	function RcptWidget() {
		$this->_order_options    = array(
			'none'     => __( 'No order', 'rcpt' ),
			'rand'     => __( 'Random order', 'rcpt' ),
			'id'       => __( 'Order by ID', 'rcpt' ),
			'author'   => __( 'Order by author', 'rcpt' ),
			'title'    => __( 'Order by title', 'rcpt' ),
			'date'     => __( 'Order by creation date', 'rcpt' ),
			'modified' => __( 'Order by last modified date', 'rcpt' ),
		);
		$this->_order_directions = array(
			'ASC'  => __( 'Ascending', 'rcpt' ),
			'DESC' => __( 'Descending', 'rcpt' ),
		);
		$widget_ops = array(
			'classname'   => 'widget_rcpt',
			'description' => __( 'Shows the most recent posts from a selected custom post type', 'rcpt' )
		);
		$this->__construct( 'rcpt', 'Recent Custom Posts Type Widget', $widget_ops );
	}

	function form( $instance ) {
		$title          = esc_attr( $instance['title'] );
		$post_type      = esc_attr( $instance['post_type'] );
		$featured_image = esc_attr( $instance['featured_image'] );
		$post_author    = esc_attr( $instance['post_author'] );
		$limit          = esc_attr( $instance['limit'] );
		$class          = esc_attr( $instance['class'] );
		$order_by       = esc_attr( $instance['order_by'] );
		$order_dir      = esc_attr( $instance['order_dir'] );

		// Fields
		$show_title      = isset( $instance['show_title'] ) ? (int) $instance['show_title'] : true; // Show by default
		$titles_as_links = isset( $instance['titles_as_links'] ) ? (int) $instance['titles_as_links'] : true; // True by default
		$show_body       = (int) $instance['show_body'];
		$show_thumbs     = esc_attr( $instance['show_thumbs'] );
		$show_dates      = esc_attr( $instance['show_dates'] );

		$fields = $instance['fields'];
		$fields = $fields ? $fields : array();

		// Set defaults
		// ...

		// Get post types
		$post_types   = get_post_types( array( 'public' => true ), 'objects' );
		$post_authors = $this->_get_post_authors();

		$html = '<p>';
		$html .= '<label for="' . $this->get_field_id( 'title' ) . '">' . __( 'Title:', 'rcpt' ) . '</label>';
		$html .= '<input type="text" name="' . $this->get_field_name( 'title' ) . '" id="' . $this->get_field_id( 'title' ) . '" class="widefat" value="' . $title . '"/>';
		$html .= '</p>';

		$html .= '<p>';
		$html .= '<label for="' . $this->get_field_id( 'post_type' ) . '">' . __( 'Post type:', 'rcpt' ) . '</label>';
		$html .= '<select name="' . $this->get_field_name( 'post_type' ) . '" id="' . $this->get_field_id( 'post_type' ) . '">';
		foreach ( $post_types as $pt ) {
			$html .= '<option value="' . $pt->name . '" ' . ( ( $pt->name == $post_type ) ? 'selected="selected"' : '' ) . '>' . $pt->label . '</option>';
		}
		$html .= '</select>';
		$html .= '</p>';

		$html .= '<p>';
		$html .= '<label for="' . $this->get_field_id( 'post_author' ) . '">' . __( 'Authored by:', 'rcpt' ) . '</label>';
		$html .= '<select name="' . $this->get_field_name( 'post_author' ) . '" id="' . $this->get_field_id( 'post_author' ) . '">';
		foreach ( $post_authors as $pa_id => $pa_name ) {
			$html .= '<option value="' . $pa_id . '" ' . ( ( $pa_id == $post_author ) ? 'selected="selected"' : '' ) . '>' . $pa_name . '&nbsp;</option>';
		}
		$html .= '</select>';
		$html .= '</p>';

		$html .= '<p>';
		$html .= '<label for="' . $this->get_field_id( 'limit' ) . '">' . __( 'Limit:', 'rcpt' ) . '</label>';
		$html .= '<select name="' . $this->get_field_name( 'limit' ) . '" id="' . $this->get_field_id( 'limit' ) . '">';
		for ( $i = 1; $i < 21; $i ++ ) {
			$html .= '<option value="' . $i . '" ' . ( ( $i == $limit ) ? 'selected="selected"' : '' ) . '>' . $i . '</option>';
		}
		$html .= '</select>';
		$html .= '</p>';

		$html .= '<p>' .
		         '<input type="checkbox" name="' . $this->get_field_name( 'featured_image' ) . '" id="' . $this->get_field_id( 'featured_image' ) . '" value="1" ' . ( $featured_image ? 'checked="checked"' : '' ) . '/>' .
		         ' <label for="' . $this->get_field_id( 'featured_image' ) . '">' . __( 'Restrict to posts with featured image', 'rcpt' ) . '</label> ' .
		         '</p>';

		$html .= '<p>';
		$html .= '<label for="' . $this->get_field_id( 'class' ) . '">' . __( 'Additional CSS class(es) <small>(optional)</small>:', 'rcpt' ) . '</label>';
		$html .= '<input type="text" name="' . $this->get_field_name( 'class' ) . '" id="' . $this->get_field_id( 'class' ) . '" class="widefat" value="' . $class . '"/>';
		$html .= '<div><small>' . __( 'One or more space separated valid CSS class names that will be applied to the generated list', 'rcpt' ) . '</small></div>';
		$html .= '</p>';

		$html .= '<p>';
		$html .= '<label for="' . $this->get_field_id( 'order_by' ) . '">' . __( 'Order by:', 'rcpt' ) . '</label>';
		$html .= '<select name="' . $this->get_field_name( 'order_by' ) . '" id="' . $this->get_field_id( 'order_by' ) . '">';
		foreach ( $this->_order_options as $key => $label ) {
			$html .= '<option value="' . $key . '" ' . ( ( $key == $order_by ) ? 'selected="selected"' : '' ) . '>' . __( $label, 'rcpt' ) . '</option>';
		}
		$html .= '</select>';
		$html .= '</p>';

		$html .= '<p>';
		$html .= '<label for="' . $this->get_field_id( 'order_dir' ) . '">' . __( 'Order direction:', 'rcpt' ) . '</label>';
		$html .= '<select name="' . $this->get_field_name( 'order_dir' ) . '" id="' . $this->get_field_id( 'order_dir' ) . '">';
		foreach ( $this->_order_directions as $key => $label ) {
			$html .= '<option value="' . $key . '" ' . ( ( $key == $order_dir ) ? 'selected="selected"' : '' ) . '>' . __( $label, 'rcpt' ) . '</option>';
		}
		$html .= '</select>';
		$html .= '</p>';

		$html .= '<p>' .
		         '<input type="checkbox" name="' . $this->get_field_name( 'show_title' ) . '" id="' . $this->get_field_id( 'show_title' ) . '" value="1" ' . ( $show_title ? 'checked="checked"' : '' ) . '/>' .
		         ' <label for="' . $this->get_field_id( 'show_title' ) . '">' . __( 'Show titles', 'rcpt' ) . '</label> ' .
		         '</p>';

		$html .= '<p>&nbsp;&nbsp;&nbsp;&nbsp;<small>' .
		         '<input type="checkbox" name="' . $this->get_field_name( 'titles_as_links' ) . '" id="' . $this->get_field_id( 'titles_as_links' ) . '" value="1" ' . ( $titles_as_links ? 'checked="checked"' : '' ) . '/>' .
		         ' <label for="' . $this->get_field_id( 'titles_as_links' ) . '">' . __( 'Titles as links to posts', 'rcpt' ) . '</label> ' .
		         '</small></p>';

		$html .= '<p>' .
		         '<input type="checkbox" name="' . $this->get_field_name( 'show_body' ) . '" id="' . $this->get_field_id( 'show_body' ) . '" value="1" ' . ( $show_body ? 'checked="checked"' : '' ) . '/>' .
		         ' <label for="' . $this->get_field_id( 'show_body' ) . '">' . __( 'Show body excerpt', 'rcpt' ) . '</label> ' .
		         '</p>';

		$html .= '<p>' .
		         '<input type="checkbox" name="' . $this->get_field_name( 'show_dates' ) . '" id="' . $this->get_field_id( 'show_dates' ) . '" value="1" ' . ( $show_dates ? 'checked="checked"' : '' ) . '/>' .
		         ' <label for="' . $this->get_field_id( 'show_dates' ) . '">' . __( 'Show post dates', 'rcpt' ) . '</label> ' .
		         '</p>';

		$html .= '<p>' .
		         '<input type="checkbox" name="' . $this->get_field_name( 'show_thumbs' ) . '" id="' . $this->get_field_id( 'show_thumbs' ) . '" value="1" ' . ( $show_thumbs ? 'checked="checked"' : '' ) . '/>' .
		         ' <label for="' . $this->get_field_id( 'show_thumbs' ) . '">' . __( 'Show featured thumbnails <small>(if available)</small>', 'rcpt' ) . '</label> ' .
		         '</p>';

		// Custom fields
		$id = sprintf( 'rcpt-custom_fields-%04d-%04d-%04d-%04d', rand(), rand(), rand(), rand() );
		$html .= '<p><a href="#toggle" class="rcpt-toggle_custom_fields" id="' . $id . '-handler">' . __( 'Show/hide custom fields', 'rcpt' ) . '</a></p>';
		$html .= '<div class="rcpt-show_custom_fields" id="' . $id . '" style="display:none">';
		$html .= '<h5>' . __( 'Custom fields', 'rcpt' ) . '</h5>';
		$html .= '<p>';
		$_fields      = $this->_get_post_fields( $post_type );
		$shown_fields = array();
		$skips        = array(
			'/^_edit_.*/',
			'/^_thumbnail_.*/',
			'/^_wp_.*/',
		);
		if ( $_fields ) {
			foreach ( $_fields as $field ) {
				if ( preg_filter( $skips, ':skip:', $field ) ) {
					continue;
				}
				$value          = in_array( $field, array_keys( $fields ) ) ? esc_attr( $fields[ $field ] ) : '';
				$shown_fields[] = '<label for="' . $this->get_field_id( 'fields' ) . '-' . $field . '">' . sprintf( __( 'Label for &quot;%s&quot;', 'rcpt' ), $field ) . '</label>' .
				                  '<input type="text" class="widefat" name="' . $this->get_field_name( 'fields' ) . '[' . $field . ']" id="' . $this->get_field_id( 'fields' ) . '-' . $field . '" value="' . $value . '" />' .
				                  '';
			}
		}
		if ( $shown_fields ) {
			$html .= '<small><em>' . __( "Fields with no associated label won't be shown in the widget output", 'rcpt' ) . '</em></small><br />';
			$html .= join( '<br />', $shown_fields );
		}
		$html .= '<small><em>' . __( 'Select your post type above and save settings to refresh', 'rcpt' ) . '</em></small>';
		$html .= '</p>';
		$html .= '</div>';
		$html .= <<<EORcptJs
<script type="text/javascript">
(function ($) {
$("#{$id}-handler").live("click", function () {
	var el = $("#{$id}");
	if (!el.length) return false;
	if (el.is(":visible")) el.hide();
	else el.show();
	return false;
});
})(jQuery);
</script>
EORcptJs;

		echo $html;
	}

	private function _get_post_fields( $type ) {
		if ( ! $type ) {
			return false;
		}
		global $wpdb;
		$fields = $wpdb->get_col(
			$wpdb->prepare( "SELECT DISTINCT meta_key FROM {$wpdb->postmeta}, {$wpdb->posts} WHERE post_id=ID and post_type='%s'", $type )
		);

		return $fields;
	}

	private function _get_post_authors() {
		global $wpdb;
		$authors = $wpdb->get_col( "SELECT DISTINCT post_author FROM {$wpdb->posts}" );
		$info    = array( '' => __( 'Anyone', 'rcpt' ) );
		foreach ( $authors as $author ) {
			$user            = new WP_User( $author );
			$info[ $author ] = $user->display_name;
		}

		return $info;
	}

	function update( $new_instance, $old_instance ) {
		$instance                   = $old_instance;
		$instance['title']          = !empty( $new_instance['title'] ) ? strip_tags( $new_instance['title'] ) : '';
		$instance['post_type']      = !empty( $new_instance['post_type'] ) ? strip_tags( $new_instance['post_type'] ) : '';
		$instance['featured_image'] = !empty( $new_instance['featured_image'] ) ? strip_tags( $new_instance['featured_image'] ) : '';
		$instance['post_author']    = !empty( $new_instance['post_author'] ) ? strip_tags( $new_instance['post_author'] ) : '';
		$instance['limit']          = !empty( $new_instance['limit'] ) ? strip_tags( $new_instance['limit'] ) : '';
		$instance['class']          = !empty( $new_instance['class'] ) ? strip_tags( $new_instance['class'] ) : '';
		$instance['order_by']       = !empty( $new_instance['order_by'] ) ? strip_tags( $new_instance['order_by'] ) : '';
		$instance['order_dir']      = !empty( $new_instance['order_dir'] ) ? strip_tags( $new_instance['order_dir'] ) : '';

		$instance['show_title']      = !empty( $new_instance['show_title'] ) ? strip_tags( $new_instance['show_title'] ) : '';
		$instance['titles_as_links'] = !empty( $new_instance['titles_as_links'] ) ? strip_tags( $new_instance['titles_as_links'] ) : '';
		$instance['show_body']       = !empty( $new_instance['show_body'] ) ? strip_tags( $new_instance['show_body'] ) : '';
		$instance['show_thumbs']     = !empty( $new_instance['show_thumbs'] ) ? strip_tags( $new_instance['show_thumbs'] ) : '';
		$instance['show_dates']      = !empty( $new_instance['show_dates'] ) ? strip_tags( $new_instance['show_dates'] ) : '';

		$instance['fields'] = array();
		$fields             = !empty( $new_instance['fields'] ) ? $new_instance['fields'] : array();
		foreach ( $fields as $key => $value ) {
			$key                        = wp_strip_all_tags( $key );
			$instance['fields'][ $key ] = wp_strip_all_tags( $value );
		}

		return $instance;
	}

	function widget( $args, $instance ) {
		extract( $args );
		$title          = apply_filters( 'widget_title', $instance['title'] );
		$post_type      = $instance['post_type'];
		$post_author    = (int) $instance['post_author'];
		$featured_image = (int) $instance['featured_image'];
		$limit          = (int) $instance['limit'];
		$class          = $instance['class'];
		$class          = $class ? " {$class}" : '';

		$order_by  = $instance['order_by'];
		$order_by  = in_array( $order_by, array_keys( $this->_order_options ) ) ? $order_by : 'none';
		$order_dir = $instance['order_dir'];
		$order_dir = in_array( $order_dir, array_keys( $this->_order_directions ) ) ? $order_dir : 'ASC';

		// Fields
		$show_title      = isset( $instance['show_title'] ) ? (int) $instance['show_title'] : true; // Show by default
		$titles_as_links = isset( $instance['titles_as_links'] ) ? (int) $instance['titles_as_links'] : true; // True by default
		$show_body       = (int) $instance['show_body'];
		$show_thumbs     = (int) $instance['show_thumbs'];
		$show_dates      = (int) $instance['show_dates'];

		$fields = $instance['fields'];
		$fields = $fields ? $fields : array();

		$query_args = array(
			'showposts'           => $limit,
			'nopaging'            => 0,
			'post_status'         => 'publish',
			'post_type'           => $post_type,
			'orderby'             => $order_by,
			'order'               => $order_dir,
			'ignore_sticky_posts' => 1
		);
		if ( $post_author ) {
			$query_args['author'] = $post_author;
		}
		if ( $featured_image ) {
			$query_args['meta_key'] = '_thumbnail_id';
		}
		$query = new WP_Query( $query_args );

		if ( $query->have_posts() ) {
			echo $before_widget;
			if ( $title ) {
				echo $before_title . $title . $after_title;
			}

			while ( $query->have_posts() ) {
				$query->the_post();

				$item_title = get_the_title() ? get_the_title() : get_the_ID();
				$image      = $src = $width = $height = false;
				if ( $show_thumbs ) {
					$thumb_id = get_post_thumbnail_id( get_the_ID() );
					if ( $thumb_id ) {
						$image = wp_get_attachment_image_src( $thumb_id, 'thumbnail' );
						if ( $image ) {
							$src    = $image[0];
							$width  = $image[1];
							$height = $image[2];
						}
					}
				}

				$image_format = $image
					? '<span class="rcpt_item_image"><img src="%s" height="%d" width="%d" alt="%s" border="0" /></span>'
					: '';
				$image_str    = sprintf( $image_format, $src, $height, $width, esc_attr( $item_title ) );

				$item_title_str       = $show_title
					? sprintf( '<span class="rcpt_item_title">%s %s</span>', $image_str, $item_title )
					: sprintf( '<span class="rcpt_item_title">%s</span>', $image_str );
				$final_item_title_str = $titles_as_links
					? sprintf( '<a href="%s" title="%s">%s</a>', get_permalink(), esc_attr( $item_title ), $item_title_str )
					: $item_title_str;

				$post_fields  = get_post_custom( get_the_ID() );
				$shown_fields = array();
				foreach ( $post_fields as $field => $value ) {
					if ( ! in_array( $field, array_keys( $fields ) ) ) {
						continue;
					} // Not here
					if ( ! $fields[ $field ] ) {
						continue;
					} // No label
					$value          = is_array( $value ) ? join( ', ', $value ) : $value;
					$shown_fields[] = array(
						'label' => wp_strip_all_tags( $fields[ $field ] ),
						'value' => wp_strip_all_tags( $value ),
					);
				}


				echo '<div class="rcpt_items"><ul class="rcpt_items_list' . $class . '">';

				echo '<li>';
				echo $final_item_title_str;
				if ( $show_body ) {
					echo '<div class="rcpt_item_excerpt">' . get_the_excerpt() . '</div>';
				}
				if ( $show_dates ) {
					echo '<span class="rcpt_item_date"><span class="rcpt_item_posted">' . __( 'Posted on', 'rcpt' ) . ' </span>' . get_the_date() . '</span>';
				}
				if ( $shown_fields ) {
					echo '<dl class="rcpt_item_custom_fields">';
					foreach ( $shown_fields as $custom ) {
						echo '<dt>' . $custom['label'] . '</dt>';
						echo '<dd>' . $custom['value'] . '</dd>';
					}
					echo '</dl>';
				}
				echo '</li>';

				echo '</ul></div>';
			}

			echo $after_widget;
		}
		wp_reset_postdata();
		wp_reset_query();
	}
}

// PHP<5.3 compatibility layer.
if ( ! function_exists( 'preg_filter' ) ) {
	function preg_filter( $pattern, $replace, $subject, $limit = - 1, $count = null ) {
		if ( ! is_array( $subject ) ) {
			$noArray = 1;
			$subject = array( $subject );
		}

		$preg = preg_replace( $pattern, $replace, $subject, $limit, $count );

		$diff = array_diff( $preg, $subject );

		if ( $noArray == 1 ) {
			$diff = implode( $diff );
		}

		return $diff;
	}
}


load_plugin_textdomain( 'rcpt', false, dirname( plugin_basename( __FILE__ ) ) . '/languages/' );

// Init widget
add_action( 'widgets_init', create_function( '', "register_widget('RcptWidget');" ) );

// Queue in the stylesheet
if ( ! is_admin() ) {
	add_action( 'init', create_function( '', 'wp_enqueue_style("rcpt_style", plugins_url("media/style.css", __FILE__));' ) );
}